ESQL: Validate unique plan attribute names

alex-spies opened the pull request:
Fix https://github.com/elastic/elasticsearch/issues/110541

There's an unwritten invariant in ES|QL that's worth spelling out: a command's/logical plan's output attributes need to have unique names. In contrast to SQL, there's no way to disambiguate in cases like

```
ROW a = 1, a = 2 | eval b = 2*a
```

because we do not have qualifiers, like `some_table.a`. Instead, ES|QL performs variable shadowing, like in `EVAL`, where the rightmost assignment wins:

```
| EVAL x = 2*some_field, y = -some_other_field, x = x + y
```

`EVAL`'s output will have the attributes `y` and `x` - the rightmost one.

Let's enforce this invariant in our dependency checker; there's only one deviation from it, currently, which is `ROW` - I think this is a bug which this PR also fixes.

This is only enforced for logical plans; the dependency checker for physical plans is yet to be enabled as part of https://github.com/elastic/elasticsearch/issues/105436.

Depends on https://github.com/elastic/elasticsearch/pull/110793

alex-spies committed:
Unique output attribute names after optimization

alex-spies committed:
Enforce unique row attribute names in verifier

alex-spies committed:
Update docs/changelog/110488.yaml

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

elasticsearchmachine commented:
Hi @alex-spies, I've created a changelog YAML for you.

alex-spies commented:
Blocked by https://github.com/elastic/elasticsearch/issues/110490

alex-spies commented on x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/analyzer/AnalyzerRules.java:
```diff
@@ -138,14 +136,6 @@ public static List<Attribute> maybeResolveAgainstList(
-                    + refs
-            )
-        );
+        throw new IllegalStateException("Reference [" + ua.qualifiedName() + "] is ambiguous; " + "matches any of " + refs);
```
Ambiguities were possible with `row` earlier; the message suggesting disambiguation does not make sense - that's not possible. This was carried over from `ql` and made sense for SQL. If we have ambiguities in ESQL, that's a bug IMO.

alex-spies commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/row.csv-spec:
```diff
@@ -69,13 +69,6 @@ a:integer | b:integer | c:null | z:integer
 ;
 
-evalRowWithNull2
-row a = 1, null, b = 2, c = null, null | eval z = a+b;
```
This one had ambiguous attribute names.

luigidellaquila replied:
Should we consider this a bug fix or a breaking change? (IMHO it's a bug, but I see it's questionable)

alex-spies replied:
I think yes, this is a bug, since there's no way to refer to the attributes named `null` in following commands.

astefan reviewed (commented):
Regarding `row` yes it seems there is an inconsistency between it and the rest of the commands. But first, let's make sure we cover all the commands from the point of view of attribute names uniqueness and if all stands then we can change `row`:
- `rename a as foo, b as foo` keeps the last `foo`
- look into `enrich`, `grok`, `dissect`, `mv_expand` regarding names uniqueness
- see if any wildcarded names (where they are allowed - `keep`, `drop` maybe some other places) don't trip the unique names verification
- combine this change with [whatever needs to change in union types](https://github.com/elastic/elasticsearch/issues/110490) so that the whole story makes sense project-wide
- look into fields and sub-fields where the name have dots - `rootfield.subfield.subsubfield` - we usually don't test these and that's a pity

alex-spies committed:
Add tests for grok, dissect, enrich

alex-spies committed:
Add tests for keep

alex-spies committed:
Make row consistent with other plans

alex-spies committed:
Update docs/changelog/110488.yaml

elasticsearchmachine commented:
Hi @alex-spies, I've updated the changelog YAML for you.

alex-spies committed:
Add test for drop, rename and stats

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into validate-unique-plan-attribute-names

astefan commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats.csv-spec:
```diff
@@ -1819,3 +1819,12 @@ warning:Line 3:17: java.lang.ArithmeticException: / by zero
+
+shadowingInternal
+FROM employees
+| STATS x = MAX(emp_no), x = MIN(emp_no)
```
I don't think we have a more extreme test, but maybe it's worth adding:

`FROM employees | STATS x = MAX(salary), x = MIN(salary) by x = gender`

alex-spies replied:
Good one, definitely worth adding.

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/OptimizerRules.java:
```diff
@@ -49,9 +54,24 @@ void checkPlan(P p, Failures failures) {
+            Set<String> outputAttributeNames = new HashSet<>();
+            Set<NameId> outputAttributeIds = new HashSet<>();
+            for (Attribute outputAttr : p.output()) {
+                if (outputAttributeNames.add(outputAttr.name()) == false || outputAttributeIds.add(outputAttr.id()) == false) {
```
`.name()` -> `.qualifiedName()`?

alex-spies replied:
Actually, I changed my mind on this:

It is the actual name that needs to be unique; otherwise, bugs could slip in because we somehow end up using qualifiers on accident; qualifiers are not respected by our optimization rules, e.g. `mergeOutputAttributes`;  this [PR](https://github.com/elastic/elasticsearch/pull/110581) demonstrates that qualifiers are entirely unused, and the validation for the current state should reflect current assumptions.

If we end up using qualifiers after all (I think that's really for the future and we should really remove them until then), we can easily update the validation.

astefan reviewed (commented):
Try also to push `mv_expand` to its edges.

alex-spies committed:
Add test dataset with deeper field hierarchy

alex-spies committed:
Add hierarchical shadowing test for eval

alex-spies committed:
Add hierarchical tests for drop, dissect

alex-spies committed:
Add hierarchical tests for enrich

alex-spies committed:
Add hierarchical tests for grok, keep

alex-spies committed:
Add hierarchical tests for rename, row

alex-spies committed:
Add more extreme case for stats

alex-spies commented:
> Regarding `row` yes it seems there is an inconsistency between it and the rest of the commands. But first, let's make sure we cover all the commands from the point of view of attribute names uniqueness and if all stands then we can change `row`:
> 
> * `rename a as foo, b as foo` keeps the last `foo`

All other commands respect unique output attribute names. Behaviorally, only `GROK` and `RENAME` stand out:
- `GROK` doesn't self-shadow; if there's 2 times the same attribute name in the pattern, like `%{WORD:foo} %{WORD:foo}`, it will create multivalues. This is consistent with the Grok ingest processor, so we want to keep it this way.
- `RENAME` renames in place, it doesn't place the renamed columns at the right hand side. For usability, I think it's important it stays this way.

> * look into `enrich`, `grok`, `dissect`, `mv_expand` regarding names uniqueness
All good here, thanks for pointing this out - let's cover all our bases!

> * see if any wildcarded names (where they are allowed - `keep`, `drop` maybe some other places) don't trip the unique names verification
Done! I was afraid about wildcards in `RENAME`, but we correctly disallow them (`VerificationException`).

> * combine this change with [whatever needs to change in union types](https://github.com/elastic/elasticsearch/issues/110490) so that the whole story makes sense project-wide

Sounds good, I'll look into it!

> * look into fields and sub-fields where the name have dots - `rootfield.subfield.subsubfield` - we usually don't test these and that's a pity

Added a new test index to our csv tests with such a hierarchical mapping, and added tests using them to all commands that are relevant IMO.

I also confirmed that we treat hierarchical fields like this like plain fields - the `name` just becomes `rootfield.subfield.subsubfield`. Qualifiers are not used. On that note, also `FieldAttribute.path` and `FieldAttribute.parent` seem unused.

alex-spies commented:
> Try also to push `mv_expand` to its edges.

Hmm, `mv_expand` takes only a single field name (resp. single attribute in the corresponding Java class); one can't even provide a list of field names due to parsing exceptions. I'm not sure which edge cases could create non unique attribute names; did you have something specific in mind?

astefan commented:
> > Try also to push `mv_expand` to its edges.
> 
> Hmm, `mv_expand` takes only a single field name (resp. single attribute in the corresponding Java class); one can't even provide a list of field names due to parsing exceptions. I'm not sure which edge cases could create non unique attribute names; did you have something specific in mind?

Nothing concrete, but mv_expand has a different/special way of defining the output (aka "expanded" attribute) by creating a mirrored attribute of the "target" one.

alex-spies committed:
Add new capability for this fix

alex-spies commented:
> > > Try also to push `mv_expand` to its edges.
> > 
> > 
> > Hmm, `mv_expand` takes only a single field name (resp. single attribute in the corresponding Java class); one can't even provide a list of field names due to parsing exceptions. I'm not sure which edge cases could create non unique attribute names; did you have something specific in mind?
> 
> Nothing concrete, but mv_expand has a different/special way of defining the output (aka "expanded" attribute) by creating a mirrored attribute of the "target" one.

I took a look; the way we're using it right now, `MvExpand` is nice and consistent.
During [resolution](https://github.com/elastic/elasticsearch/blob/bfc32a2acc8942dea31c7ac1f893805906b833fb/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java#L493-L507), we hand it a new reference attribute (not referring to anything, similar hack to enrich) with the same name.

When computing the [output](https://github.com/elastic/elasticsearch/blob/bfc32a2acc8942dea31c7ac1f893805906b833fb/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/MvExpand.java#L34-L38), we replace the initial, multi-valued attribute by this new reference attribute with the same name.

This is not failing the unique name validation added in this PR, so I think we should be good.

The way I can see MV_EXPAND compute inconsistent output, is only if the child plan before `MV_EXPAND` is already inconsistent (multiple attributes with the name of the expanded attr), or we misuse `MV_EXPAND` by giving it a differently named `expanded` field (currently doesn't happen in the LogicalPlanBuilder), or an optimization is buggy.

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into validate-unique-plan-attribute-names

alex-spies committed:
Fix EsRelation.equals, mutation in ResolveUnionTypes

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into validate-unique-plan-attribute-names

alex-spies committed:
Make union types use unique attribute names

alex-spies committed:
Cleanup leftover

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into validate-unique-plan-attribute-names

alex-spies committed:
Revert "Cleanup leftover"

This reverts commit 658bebcdc8f802cc034e31c9607150ef76960a7e.

alex-spies committed:
Revert "Make union types use unique attribute names"

This reverts commit 1a648e4a05d3c6172d2581b678bfa7ceee6176a3.

alex-spies committed:
Revert "Fix EsRelation.equals, mutation in ResolveUnionTypes"

This reverts commit 2cb23f594dcf0b1a0d1d9f96b5a7386b3e97fcd3.

alex-spies committed:
More ENRICH tests with internal shadowing

alex-spies committed:
More consistent test names

alex-spies committed:
More KEEP tests

alex-spies committed:
Update docs

alex-spies committed:
Add more tests

alex-spies committed:
Improve doc wording

alex-spies committed:
Improve GROK docs

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into validate-unique-plan-attribute-names

alex-spies commented:
Okay, this will continue failing CI until https://github.com/elastic/elasticsearch/pull/110793 is merged, but otherwise should be ready.

leemthompo commented on docs/reference/esql/processing-commands/enrich.asciidoc:
```diff
@@ -31,11 +31,16 @@ name as the `match_field` defined in the <<esql-enrich-policy,enrich policy>>.
 existing column will be replaced by the new column. If not specified, each of
-the enrich fields defined in the policy is added
+the enrich fields defined in the policy is added.
+If a column has the same name as the enrich field, it will be discarded unless
+that field is given a new name.
```
```suggestion
A column with the same name as the enrich field will be discarded unless 
the enrich field is renamed.
```

leemthompo commented on docs/reference/esql/processing-commands/grok.asciidoc:
```diff
@@ -20,6 +20,9 @@ multiple values, `GROK` will process each value.
 
 `pattern`::
 A grok pattern.
+In case a field name coincides with an existing column, the existing column is discarded.
```
```suggestion
If a field name conflicts with an existing column, the existing column is discarded.
```

leemthompo commented on docs/reference/esql/processing-commands/grok.asciidoc:
```diff
@@ -67,4 +70,16 @@ include::{esql-specs}/docs.csv-spec[tag=grokWithToDatetime]
 include::{esql-specs}/docs.csv-spec[tag=grokWithToDatetime-result]
 |===
+
+In case a field name is used more than once, `GROK` creates a multi-valued
```
```suggestion
If a field name is used more than once, `GROK` creates a multi-valued
```

leemthompo commented on docs/reference/esql/processing-commands/keep.asciidoc:
```diff
@@ -16,6 +16,10 @@ KEEP columns
 
 `columns`::
 A comma-separated list of columns to keep. Supports wildcards.
+In case a column name without wildcards occurs multiple times, the column is placed at the
+position of the rightmost duplicate. The same is true if a column name matches multiple
+wildcards. If a column name matches both a wildcard and also a column name without wildcards,
+the position of the latter is used.
```
```suggestion
If a column name without wildcards occurs multiple times, the column is placed at the
position of the rightmost duplicate. The same is true if a column name matches multiple
wildcards. If a column name matches both a wildcard and also a column name without wildcards,
the position of the latter is used.
```

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into validate-unique-plan-attribute-names

leemthompo commented on docs/reference/esql/processing-commands/lookup.asciidoc:
```diff
@@ -18,6 +18,8 @@ LOOKUP table ON match_field1[, match_field2, ...]
 
 `table`::
 The name of the `table` provided in the request to match.
+If it has column names that conflict with existing columns, the table's columns replace the
```
```suggestion
If the table's column names conflict with existing columns, the table's columns replace the
```

leemthompo reviewed (commented):
Few very minor docs edits for brevity and consistency. :)

Otherwise we use both "discarded" and "dropped" here, maybe "dropped" should be used throughout?

alex-spies committed:
Update RENAME docs and tests

alex-spies committed:
Avoid duplicate field attribs from union type res

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -1068,21 +1068,37 @@ public static Expression castStringLiteral(Expression from, DataType target) {
+                }
+            });
+
+            return plan.transformUp(LogicalPlan.class, p -> p.resolved() || p.childrenResolved() == false ? p : doRule(p));
```
Not sure I understand why p should be resolved, while p.children should be unresolved. Should it not be `p -> p.resolved() ==false || p.childrenResolved() == false ? p : doRule(p)`

alex-spies replied:
This was carried over from `AnalyzerRule` and `BaseAnalyzerRule` while I collapse their execution logic into this simple `Rule`.

Currently it says "I'll just return the current plan without changes (i.e. skip) if it's either already resolved or has yet-to-be resolved children". I think this is correct.

astefan commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/rename.csv-spec:
```diff
@@ -174,3 +174,42 @@ avg_worked_seconds:l | birth_date:date          | emp_no:i             | first_n
+FROM employees
+| SORT emp_no ASC
+| KEEP emp_no, first_name, last_name
+| RENAME emp_no AS last_name
```
That's an interesting edge case.
Maybe I am reading the docs update wrong (thank you for updating the docs as well):

> If it conflicts with an existing column name,
> the existing column is replaced by the renamed column. If multiple columns
> are renamed to the same name, all but the rightmost column with the same new
> name are dropped.

but there is a contradiction between what we say in docs and what actually happens. According to docs, in the results below, `last_name` should contain info from `emp_no` but it should be the second column from left to right, not the first one.

alex-spies committed:
Fix leftovers

alex-spies committed:
Make tests deterministic

alex-spies committed:
Fix rename shadowing docs

alex-spies commented on docs/reference/esql/processing-commands/eval.asciidoc:
```diff
@@ -16,10 +16,12 @@ EVAL [column1 =] value1[, ..., [columnN =] valueN]
 
 `columnX`::
 The column name.
+In case a column name coincides with an existing column, the existing column is discarded.
```
I'll go for `If a column with the same name already exists, the existing column is dropped.` if that works for you.

alex-spies committed:
Apply Liam's doc remarks

astefan reviewed (approved):
LGTM

luigidellaquila reviewed (approved):
LGTM, thanks!

alex-spies committed:
Don't describe KEEP precedence twice

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into validate-unique-plan-attribute-names

alex-spies commented:
Thanks for your reviews, @leemthompo , @astefan , @craigtaverner and @luigidellaquila !

I updated `keep.asciidoc` one more time (I noticed the info I added would be duplicated), then it's time for a merge!

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  8.15  | Commit could not be cherrypicked due to conflicts |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 110488`

alex-spies commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.15|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/110966">](https://github.com/elastic/elasticsearch/pull/110966)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 9.4.3 -->
